### DISCUSSION TITLE
Event triggered conditional preferences profiles - electric carbon intensity, variable rate adjustment, etc

### DISCUSSION THREAD

**Discussion by brierjon** [BOINC/boinc#5215] [Ideas]
While the objective of BOINC is to harness idle computing power for science and other projects, it doesn't mean that there couldn't or shouldn't be considerations of energy use and climate affects of the increased computing load. 

A few concepts I've wondered if BOINC devs have considered or discussed:

1. What do we know of the carbon intensity and electricity makeup of BOINC systems? If so, please post a link to something published.
2. Would an option to set computing profiles to minimize the carbon footprint or time of day/variable electric rates increase the sustainability or lower costs of participanting BOINC computing?
3. How does BOINC distributed computing compare to other computing services in sustainable terms?  
4. Could leveraging a current rate or predictive carbon intensity API like https://www.co2signal.com/ allow BOINC systems to adjust how they set their system compute performance? 
5. If there are sufficient number of systems in a grid region would scaling back computing resources while carbon intensity is high or scaling up when renewables are high make BOINC more sustainable? 

I think the concept of sustainable measures and event triggered preferences should be able to leverage much of the existing preferences controls and simply adjusting # of cores or % processing power for a period of time. This may require adjustment on how BOINC clients and account managers integrate and coordinate for preferences control. 

Having carbon and cost aware computing could also open the opportunity to measure BOINC's footprint to compare to other computing approaches and automate costs controls where electric rates are variable or have peak hours.

While just an example of a possible free API for non-commercial use to consider: https://www.co2signal.com/ is an API available in Home Assistant (https://www.home-assistant.io/integrations/co2signal/)

There are probably other cases that an event triggered preference mechanism might be valuable (or it may be possible now and not clearly communicated or leveraged). This may not be a high priority for resource allocation, but if a researcher wants to take this idea and run with it. Please do!

**Comment by davidpanderson** [comment]
There's a proposal for a "generalized" prefs system that would integrate info from external sources:
https://github.com/BOINC/boinc/wiki/Prefs2
However, this would be a large change and is unlikely to happen soon.

A more practical (but less general) approach would be to use a separate program
that monitors power parameters and turns BOINC computing on/off either using RPCs:
https://github.com/BOINC/boinc/wiki/GuiRpc
or using boinccmd:
https://boinc.berkeley.edu/wiki/Boinccmd_tool
